tools/cellxgene_census_builder/src/cellxgene_census_builder/build_soma/anndata.py

```diff
@@ -19,99 +23,244 @@
+      only by an index (either a Python range, or a numpy point index array)
+
+    This effectively:
+    * removes the overhead of reading unused slots such as obsm/varm/obsp/varp
+    * removes the overhead when you need to index a view (where AnnData materializes)
+
+    In the future, if we need additional slots accessible, we may need to add sub-proxies
+    for uns, obs*, var*, etc. But at the moment, the Builder has no use for these so they
+    are omitted.
+    """
+
+    _obs: pd.DataFrame
+    _var: pd.DataFrame
+    _X: h5py.Dataset | CSRDataset | CSCDataset
+
+    def __init__(
+        self,
+        filename: str | PathLike[str],
+        *,
+        view_of: Self | None = None,
+        obs_idx: slice | npt.NDArray[np.int64] | None = None,
+        var_idx: slice | npt.NDArray[np.int64] | None = None,
+        obs_column_names: Optional[Tuple[str, ...]] = None,
+        var_column_names: Optional[Tuple[str, ...]] = None,
+    ):
+        self.filename = filename
+        self.is_view = view_of is not None
+
+        if view_of is None:
+            self._obs, self._var, self._X = self._load_h5ad(obs_column_names, var_column_names)
+            self._obs_idx: slice | npt.NDArray[np.int64] = slice(None)
+            self._var_idx: slice | npt.NDArray[np.int64] = slice(None)
+        else:
+            self._obs, self._var, self._X = (view_of._obs, view_of._var, view_of._X)
+            assert obs_idx is not None
+            assert var_idx is not None
+            self._obs_idx = obs_idx
+            self._var_idx = var_idx
+
+    @property
+    def X(self) -> sparse.spmatrix | npt.NDArray[np.integer[Any] | np.floating[Any]]:
+        # Do two separate slices, as the underlying AnnData CS*Dataset
+        # does bad things when you do slicing on both axis simultaneously.
+        # Rely on the fact that we need to be most selective on the obs
+        # dimension, and let scipy.sparse handle the second slice.
+        X = self._X[self._obs_idx][:, self._var_idx]
+        if sparse.isspmatrix(X) and not X.has_canonical_format:
             X.sum_duplicates()
+        return X
+
+    @cached_property
+    def obs(self) -> pd.DataFrame:
+        return self._obs.iloc[self._obs_idx]
+
+    @cached_property
+    def var(self) -> pd.DataFrame:
+        return self._var.iloc[self._var_idx]
+
+    @property
+    def n_obs(self) -> int:
+        if isinstance(self._obs_idx, slice):
+            return len(range(*self._obs_idx.indices(len(self._obs.index))))
+        return len(self.obs)
+
+    @property
+    def n_vars(self) -> int:
+        if isinstance(self._var_idx, slice):
+            return len(range(*self._var_idx.indices(len(self._var.index))))
+        return len(self.var)
+
+    @property
+    def shape(self) -> tuple[int, int]:
+        return self.n_obs, self.n_vars
+
+    def __len__(self) -> int:
+        return self.shape[0]
+
+    def __getitem__(self, key: Index) -> "AnnDataProxy":
```

Comment:
This method and/or _index_index could probably use some explicit tests. It handles lots cases like stepped slices, using slices vs masks vs points, etc..  If nothing else, the tests will help document the explicit requirements of how an anndataproxy view can be sliced.  Maybe also a comment, "supports slicing in the same manner as AnnData"

I might rename `_index_index` to `_slice_index`, though naming is a personal matter. :) AnnData docs do refer to this operation "slicing", so seems reasonable.